knode/knode.go

```diff
@@ -0,0 +1,68 @@
+package knode
+
+import (
+	"errors"
+	"os"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+var Kubeconfig *string
+
+type K8sNotifier struct {
+	nodeName string
+	config   *rest.Config
+	recorder record.EventRecorder
+}
+
+func NewK8sNotifier() *K8sNotifier {
+	return &K8sNotifier{}
+}
+
+// InitializeK8sNotifier sets up the event recorder to send events to the Kube API
+// server.
+func (k *K8sNotifier) InitializeK8sNotifier() error {
+	var err error
+	if k.config, err = config.GetConfig(); err != nil {
+		klog.Fatalf("cannot retrieve rest config for kubernetes. %v ", err)
+	}
+	// Proceed to fetch the NODE_NAME set through the downward API.
+	k.nodeName = os.Getenv("NODE_NAME")
+	// If the node name is still unset, obtain the hostname as it may be running in standalone
+	// mode.
+	if k.nodeName == "" {
+		k.nodeName, err = os.Hostname()
+		if err != nil || k.nodeName == "" {
+			return errors.New("hostname is unavailable, node events cannot be reported")
+		}
+	}
+
+	client := kubernetes.NewForConfigOrDie(k.config).CoreV1()
+	k.recorder = getEventRecorder(client, "", k.nodeName, "RTASNotifier")
+	return nil
+}
+
+func getEventRecorder(c typedcorev1.CoreV1Interface, namespace, nodeName, source string) record.EventRecorder {
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(legacyscheme.Scheme, v1.EventSource{Component: source, Host: nodeName})
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: c.Events(namespace)})
+	return recorder
+}
+
+// NotifyAPIServer posts events to the Kubernetes API server.
+func (k *K8sNotifier) NotifyAPIServer(message string) {
+	k.recorder.Event(&v1.ObjectReference{
+		Kind: "Node",
+		Name: k.nodeName,
+		UID:  types.UID(k.nodeName),
+	}, v1.EventTypeWarning, "PlatformEvent", message)
```

Comment:
why `v1.EventTypeWarning` is always warning? can we set the severity type based on the type of events we are getting from the DB?

Reply:
Agree, we can cascade the severity from the RTAS log, I was trying to keep it inline with the pre-defined types of events in K8s. 
Should it be okay to deviate from the existing type? 
https://github.com/kubernetes/api/blob/ba9d27cc9c7de0882d39035d8ae3b18d1434b636/core/v1/types.go#L6159-L6165

Reply:
na, what I really meant is - we can map the events from RTAS logs to k8s event types

Reply:
Got your point.. But I still have some confusion.

The defined k8s event types are only Warning and Normal. While there may be events from other components with the existing type, the RTAS events are of 7 different levels - Debug, Info, Event, Warning, Error_local, Error and Fatal. While we can Map Warning to warning and Debug and Info to Normal events, the types Error_local, Error and Fatal don't have a corresponding severity type defined in core-types..

My doubt is whether it'd be okay to have new types of severity appearing under events( Like Error, Fatal, etc.), while the available Kubernetes defined types of events are only Normal and Warning..

Reply:
In that case let's map to whatever levels we have, like serious events as warning and not so serious issue as Info